planner/cascades/transformation_rules.go

```diff
@@ -2502,3 +2505,133 @@ func (r *PullSelectionUpApply) OnTransform(old *memo.ExprIter) (newExprs []*memo
 	newApplyGroupExpr.SetChildren(outerChildGroup, old.Children[1].GetExpr().Children[0])
 	return []*memo.GroupExpr{newApplyGroupExpr}, false, false, nil
 }
+
+// MergeAdjacentWindow merge adjacent Window.
+type MergeAdjacentWindow struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentWindow creates a new Transformation MergeAdjacentWindow.
+// The pattern of this rule is `Window -> Window`.
+func NewRuleMergeAdjacentWindow() Transformation {
+	rule := &MergeAdjacentWindow{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandWindow,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandWindow, memo.EngineAll),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *MergeAdjacentWindow) Match(expr *memo.ExprIter) bool {
+	curWinPlan := expr.GetExpr().ExprNode.(*plannercore.LogicalWindow)
+	nextGroupExpr := expr.Children[0].GetExpr()
+	nextWinPlan := nextGroupExpr.ExprNode.(*plannercore.LogicalWindow)
+	nextSchema := nextGroupExpr.Group.Prop.Schema
+	nextGroupChildren := nextGroupExpr.Children
+	ctx := expr.GetExpr().ExprNode.SCtx()
+
+	// Whether Partition parts are the same.
+	if len(curWinPlan.PartitionBy) != len(nextWinPlan.PartitionBy) ||
+		len(curWinPlan.OrderBy) != len(nextWinPlan.OrderBy) ||
+		(curWinPlan.Frame == nil && nextWinPlan.Frame != nil) ||
+		(curWinPlan.Frame != nil && nextWinPlan.Frame == nil) {
+		return false
+	}
+	for i, item := range curWinPlan.PartitionBy {
+		if !item.Col.Equal(ctx, nextWinPlan.PartitionBy[i].Col) ||
+			item.Desc != nextWinPlan.PartitionBy[i].Desc {
+			return false
+		}
+	}
+
+	// Whether OrderBy parts are the same.
+	for i, item := range curWinPlan.OrderBy {
+		if !item.Col.Equal(ctx, nextWinPlan.OrderBy[i].Col) ||
+			item.Desc != nextWinPlan.OrderBy[i].Desc {
+			return false
+		}
+	}
+
+	// Whether Frame parts are the same.
+	if curWinPlan.Frame != nil {
+		if curWinPlan.Frame.Type != nextWinPlan.Frame.Type ||
+			curWinPlan.Frame.Start.Type != nextWinPlan.Frame.Start.Type ||
+			curWinPlan.Frame.Start.UnBounded != nextWinPlan.Frame.Start.UnBounded ||
+			curWinPlan.Frame.Start.Num != nextWinPlan.Frame.Start.Num ||
+			curWinPlan.Frame.End.Type != nextWinPlan.Frame.End.Type ||
+			curWinPlan.Frame.End.UnBounded != nextWinPlan.Frame.End.UnBounded ||
+			curWinPlan.Frame.End.Num != nextWinPlan.Frame.End.Num {
+			return false
+		}
+		for i, expr := range curWinPlan.Frame.Start.CalcFuncs {
+			if !expr.Equal(ctx, nextWinPlan.Frame.Start.CalcFuncs[i]) {
+				return false
+			}
+		}
+		for i, expr := range curWinPlan.Frame.End.CalcFuncs {
+			if !expr.Equal(ctx, nextWinPlan.Frame.End.CalcFuncs[i]) {
+				return false
+			}
```

Comment:
Maybe we could implement an `Equal` function for the `FrameBound` structure,  it will simplify the code here.